security/pkg/pki/ca/selfsignedcarootcertrotator.go

```diff
@@ -0,0 +1,232 @@
+		}
+		return
+	}
+
+	rootCertRotatorLog.Infof("Refresh root certificate, root cert is about to expire: %s", err.Error())
+	options := util.CertOptions{
+		TTL:           rotator.config.caCertTTL,
+		SignerPrivPem: caSecret.Data[caPrivateKeyID],
+		Org:           rotator.config.org,
+		IsCA:          true,
+		IsSelfSigned:  true,
+		RSAKeySize:    caKeySize,
+		IsDualUse:     rotator.config.dualUse,
+	}
+	pemCert, pemKey, ckErr := util.GenRootCertFromExistingKey(options)
+	if ckErr != nil {
+		rootCertRotatorLog.Errorf("unable to generate CA cert and key for self-signed CA: %s", ckErr.Error())
+		return
+	}
+
+	oldCaCert := caSecret.Data[caCertID]
+	oldCaPrivateKey := caSecret.Data[caPrivateKeyID]
+	if rollback, err := rotator.updateRootCertificate(caSecret, true, pemCert, pemKey); err != nil {
+		if !rollback {
+			rootCertRotatorLog.Errorf("Failed to roll forward root certificate (error: %s). "+
+				"Abort new root certificate", err.Error())
+			return
+		}
+		// caSecret is out-of-date. Need to load the latest istio-ca-secret to roll back root certificate.
+		_, err = rotator.updateRootCertificate(nil, false, oldCaCert, oldCaPrivateKey)
+		if err != nil {
+			rootCertRotatorLog.Errorf("Failed to roll backward root certificate (error: %s).", err.Error())
+		}
+		return
+	}
+	rootCertRotatorLog.Info("Root certificate rotation is completed successfully.")
+}
+
+// updateRootCertificate updates root certificate in istio-ca-secret, keycertbundle and configmap. It takes a scrt
+// object, cert, and key, and a flag rollForward indicating whether this update is to roll forward root certificate or
+// to roll backward.
+// updateRootCertificate returns error when any step is failed, and a flag indicating whether a rollback is required.
+// Only when rollForward is true and failure happens, the returned rollback flag is true.
+func (rotator *SelfSignedCARootCertRotator) updateRootCertificate(caSecret *v1.Secret, rollForward bool, cert, key []byte) (bool, error) {
+	var err error
+	if caSecret == nil {
+		caSecret, err = rotator.caSecretController.LoadCASecretWithRetry(CASecret,
+			rotator.config.caStorageNamespace, rotator.config.retryInterval, 30*time.Second)
+		if err != nil {
+			return false, fmt.Errorf("failed to load CA secret %s:%s (error: %s)", rotator.config.caStorageNamespace, CASecret,
+				err.Error())
+		}
+	}
+	caSecret.Data[caCertID] = cert
+	caSecret.Data[caPrivateKeyID] = key
+	if err = rotator.caSecretController.UpdateCASecretWithRetry(caSecret, rotator.config.retryInterval, 30*time.Second); err != nil {
+		// CA keycertbundle is unlikely to fail, no need to roll back.
+		return false, fmt.Errorf("failed to update CA secret (error: %s)", err.Error())
+	}
+	rootCertRotatorLog.Infof("Root certificate is written into CA secret: %v", string(cert))
+	if err := rotator.ca.GetCAKeyCertBundle().VerifyAndSetAll(cert, key, nil, cert); err != nil {
+		return false, fmt.Errorf("failed to update CA KeyCertBundle (error: %s)", err.Error())
```

Comment:
I think you still need to return true here, because the root secret and the config map are inconsistent.